hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieSparkSqlWriter.scala

```diff
@@ -1063,7 +1063,9 @@ object HoodieSparkSqlWriter {
     val recordType = config.getRecordMerger.getRecordType
 
     val shouldCombine = parameters(INSERT_DROP_DUPS.key()).toBoolean ||
-      operation.equals(WriteOperationType.UPSERT) ||
+      (operation.equals(WriteOperationType.UPSERT) &&
+        parameters.getOrElse(HoodieWriteConfig.COMBINE_BEFORE_UPSERT.key(),
+          HoodieWriteConfig.COMBINE_BEFORE_UPSERT.defaultValue()).toBoolean) ||
```

Comment:
This also makes spark hudi compatible with flink hudi, if COMBINE_BEFORE_UPSERT=false, users can upsert when no preCombine field is not defined. Flink has COMBINE_BEFORE_UPSERT set to false as default. 
That's why it worked in flink but not in spark.

Reply:
Actually only MOR table can allow lossening the restriction for COMBINE_BEFORE_UPSERT, for COW table, the combining of in-mem dataset is a precondition for the uniqueness of deduplication, Flink does that in this line:
https://github.com/apache/hudi/blob/58e9ce710b405e7653236989912eae49559cd011/hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/HoodieTableFactory.java#L382

Reply:
Does this mean that users should never set COMBINE_BEFORE_UPSERT for CoW table? I thought this config is something that users are allowed to disable deliberately.

If combine before upsert is always required for CoW, how Flink deduplicates records for no precombine tables?

My thinking was:
1. user creates table without precombine field
2. wants to do update
3. can't because it's getting exception about missing precombine field

To allow this flow:
1. user creates table with no precombine field
2. hudi internally knows no precombine filed was defined and disables COMBINE_BEFORE_UPSERT
3. Hudi does not try to precombine on upsert, it's user responsibility to deduplicate incoming dataset

Thought this was a clean way to do it.

Reply:
> Thought this was a clean way to do it.

We can do some config inference here which is good way to go, but as I said, for COW table, the in-mem combining is a must, you can not break that rule.

Reply:
For case where the source is a hudi table or some online table (with uniqueness of rows guaranteed), the source batch is guaranteed to be de-duped. In this case, isn't it safe to allow users to disable pre-combining independent of whether this is COR or MOR. @danny0405 : Let me know if I am missing something ? Thanks

Reply:
Yeah, we uniqueness can be assured by the upstream data source, then the user should use the insert operation instead of the upsert operation right? It is risky to loosen the restriction if there are no good knowledge for user to learn about these details.

We can loosen the restriction if we have a good way to tell user there is risky for data duplications.